Allow testing of commands that return multiple messages with respond_with_slack_message

gcraig99 opened the pull request:
Allows testing of a single response out of multiple responses, and also allows testing of all responses from a single command call. 

Resolves #135

dblock commented:
I like the idea. This needs tests and a passing build, please. I would use "once" instead of "at least once" to have a stronger expectation. Maybe something more elaborate can afford for multiple repeated messages.

dblock commented:
I also wonder whether we should instead introduce `to_respond_with_slack_messages` (plural)?

gcraig99 commented:
Refactored to update support for a single message from a group of messages using respond_with_slack_message and added new respond_with_slack_messages to allow checking of multiple responses at once. Agree that it's cleaner that way.  Added tests for both matchers and updated README.md with notes on new matcher. A single test case is still failing for me, but it appears to be unrelated to any of my work. The VCR is not matching the response for migration_in_progress and is failing with an error that no response is available in the VCR.

gcraig99 commented:
Build is still failing due to the above mentioned test case. If I'm missing something just let me know, but it doesn't appear that I'm the cause of that failure, copied below. 
1) SlackRubyBot::Server retries on rtm.start errors migration_in_progress
     Failure/Error:
       expect do
         subject.run
       end.to raise_error Slack::Web::Api::Error, 'unknown'
     
       expected Slack::Web::Api::Errors::SlackError with "unknown", got #<VCR::Errors::UnhandledHTTPRequestError:

dblock commented on CHANGELOG.md:
```diff
@@ -1,6 +1,7 @@
 ### 0.10.5 (Next)
 
 * Refactored `SlackRubyBot::MVC::Controller::Base`, consolidated ivar handling, centralized object allocations and DRYed up the code - [@chuckremes](https://github.com/chuckremes).
+* [#157](https://github.com/slack-ruby/slack-ruby-bot/pull/157): Added support for checking multiple responses at once via new `respond_with_slack_messages` matcher, and single response out of multiple responses to `response_with_slack_message` matcher - [@gcraig99](https://github.com/gcraig99) 
```
We already had `respond_with_slack_message` (typo btw), so I would just say "Added `respond_with_slack_messages` RSpec expectation".

You need a period at the end of this for Danger to be happy ;)

dblock commented on README.md:
```diff
@@ -547,6 +547,7 @@ Slack-ruby-bot ships with a number of shared RSpec behaviors that can be used in
 
 * [behaves like a slack bot](lib/slack-ruby-bot/rspec/support/slack-ruby-bot/it_behaves_like_a_slack_bot.rb): A bot quacks like a Slack Ruby bot.
 * [respond with slack message](lib/slack-ruby-bot/rspec/support/slack-ruby-bot/respond_with_slack_message.rb): The bot responds with a message.
+* [respond with slack messages](lib/slack-ruby-bot/rspec/support/slack-ruby-bot/respond_with_slack_messages.rb): The bot responds with a set of message.
```
set of messageS or just "multiple messages"

dblock commented on lib/slack-ruby-bot/rspec/support/slack-ruby-bot/respond_with_slack_message.rb:
```diff
@@ -13,8 +13,9 @@
-    expect(client).to receive(:message).with(channel: channel, text: expected)
+    allow(client).to receive(:message)
     message_command.call(client, Hashie::Mash.new(text: message, channel: channel, user: user))
+    expect(client).to have_received(:message).with(channel: channel, text: expected).once
```
I believe this changes behavior, ie. it allows the client to receive multiple messages that aren't the one expected plus the one expected. I think we don't want that, we want expectations to be "precise".

dblock commented on lib/slack-ruby-bot/rspec/support/slack-ruby-bot/respond_with_slack_messages.rb:
```diff
@@ -0,0 +1,30 @@
+require 'rspec/expectations'
+RSpec::Matchers.define :respond_with_slack_messages do |expected|
+  match do |actual|
+    raise ArgumentError, 'respond_with_slack_messages expects an array of ordered responses' unless expected.is_a? Array
```
I would check for `Enumerable` instead or maybe just that it responds to `each`? This way I can get the messages from some collection, database, whatever, if needed.

dblock commented on lib/slack-ruby-bot/rspec/support/slack-ruby-bot/respond_with_slack_messages.rb:
```diff
@@ -0,0 +1,30 @@
+    expected.each do |exp|
+      expect(client).to have_received(:message).with(channel: channel, text: exp).once
+    end
+    true
```
Same as above, I think we want exact expectations, making sure only the messages expected have been received. This behavior should have tests in both cases.

dblock commented on lib/slack-ruby-bot/rspec/support/slack-ruby-bot/respond_with_slack_messages.rb:
```diff
@@ -0,0 +1,30 @@
+  def parse(actual)
+    actual = { message: actual } unless actual.is_a?(Hash)
+    [actual[:channel] || 'channel', actual[:user] || 'user', actual[:message]]
+  end
```
There's some duplication here with the other expectation. Maybe put the common code in a module and include it?

dblock commented:
Without debugging, the VCR error means that something is making an HTTP request that wasn't expecting. You can record that and look at what's being made by changing vcr.rb to record new episodes, but most likely there's a behavior change in the code if it's failing an old spec.

gcraig99 commented:
Ok, I think I've addressed all of the concerns and we have a passing build now. If there's anything else just let me know.

dblock commented on lib/slack-ruby-bot/rspec/support/slack-ruby-bot/respond_with_slack_message.rb:
```diff
@@ -1,27 +1,32 @@
+
+    def client.say(options = {})
+      super
+      @test_received_messages = @test_received_messages.nil? ? '' : @test_received_messages
```
Sending messages with `\n` is going to confuse things, so make this an array:

```ruby
test_received_messages ||= []
test_received_messages << options
```

dblock commented on lib/slack-ruby-bot/rspec/support/slack-ruby-bot/respond_with_slack_message.rb:
```diff
@@ -1,27 +1,32 @@
-               SlackRubyBot::Client.new
-             end
+    client = respond_to?(:client) ? send(:client) : SlackRubyBot::Client.new
+    def client.test_messages
```
Try `client.attr_accessor :test_messages` instead?

gcraig99 replied:
No go, that causes an error since it's an instance of a class, rather than a class definition.

dblock commented on lib/slack-ruby-bot/rspec/support/slack-ruby-bot/respond_with_slack_message.rb:
```diff
@@ -1,27 +1,32 @@
-    expect(client).to receive(:message).with(channel: channel, text: expected)
+    allow(client).to receive(:message)
     message_command.call(client, Hashie::Mash.new(text: message, channel: channel, user: user))
+    @messages = client.test_messages
```
Why not leave the old code with `expect`, and then you don't need the `failure_message` at all?

dblock commented on lib/slack-ruby-bot/rspec/support/slack-ruby-bot/respond_with_slack_messages.rb:
```diff
@@ -0,0 +1,34 @@
+    allow(client).to receive(:message)
+    message_command.call(client, Hashie::Mash.new(text: message, channel: channel, user: user))
+    @messages = client.test_messages
+    expected.each do |exp|
```
If you make this `expect(...)` above you don't need to do this whole saving of `@messages` and `failure_message`, no? This seems unnecessarily complicated.

gcraig99 replied:
The expect statements are still there, the issue is that when using the expect statement without the custom failure message and the storage of the received messages, you don't get any useful output. 

#### This is the output without the custom failure message:
 1) RSpec it_fails_on_purpose_to_show_output
     Failure/Error:
       expect(message: "#{SlackRubyBot.config.user} single message")
         .to respond_with_slack_message(/failure message/i)
     
       `expected {:message=>"rubybot single message"} to respond with slack message /failure message/i`
 
  2) RSpec respond_with_failure_to_show_output
     Failure/Error:
       expect(message: "#{SlackRubyBot.config.user} respond 5 times")
         .to respond_with_slack_messages(expected_responses)
     
       `expected {:message=>"rubybot respond 5 times"} to respond with slack messages /response 0/i, /response 1/i, /response 2/i, /response 3/i, /response 4/i, and /response 7/i`

#### This is the output with the custom failure message, much more useful in figuring out why it failed.     
1) RSpec it_fails_on_purpose_to_show_output
     Failure/Error:
       expect(message: "#{SlackRubyBot.config.user} single message")
         .to respond_with_slack_message(/failure message/i)
     
       `expected to receive message with text: (?i-mx:failure message) once,
        received:[{:text=>"single response", :channel=>"channel"}]`

  2) RSpec respond_with_failure_to_show_output
     Failure/Error:
       expect(message: "#{SlackRubyBot.config.user} respond 5 times")
         .to respond_with_slack_messages(expected_responses)
     
       `expected to receive messages in order with text: [/response 0/i, /response 1/i, /response 2/i, /response 3/i, /response 4/i, /response 7/i]
        received:[{:text=>"response 0", :channel=>"channel"}, {:text=>"response 1", :channel=>"channel"}, {:text=>"response 2", :channel=>"channel"}, {:text=>"response 3", :channel=>"channel"}, {:text=>"response 4", :channel=>"channel"}]`

dblock commented:
Merging, thanks for the good work.
